core/chains/evm/gas/block_history_estimator.go

```diff
@@ -187,53 +223,151 @@ func (b *BlockHistoryEstimator) GetLegacyGas(_ []byte, gasLimit uint32, maxGasPr
+		return nil, 0, errors.Errorf("GasPrice unset. Likely cause is that BlockHistoryEstimator failed to make the first estimation on start, or no suitable transactions are available for estimation (size=%d)", b.size)
 	}
 	gasPrice = capGasPrice(gasPrice, maxGasPriceWei, b.config)
 	return
 }
 
-func (b *BlockHistoryEstimator) getGasPrice() *big.Int {
-	b.mu.RLock()
-	defer b.mu.RUnlock()
+func (b *BlockHistoryEstimator) getGasPrice() *assets.Wei {
+	b.priceMu.RLock()
+	defer b.priceMu.RUnlock()
 	return b.gasPrice
 }
-func (b *BlockHistoryEstimator) getTipCap() *big.Int {
-	b.mu.RLock()
-	defer b.mu.RUnlock()
+func (b *BlockHistoryEstimator) getTipCap() *assets.Wei {
+	b.priceMu.RLock()
+	defer b.priceMu.RUnlock()
 	return b.tipCap
 }
 
-func (b *BlockHistoryEstimator) BumpLegacyGas(originalGasPrice *big.Int, gasLimit uint32, maxGasPriceWei *big.Int) (bumpedGasPrice *big.Int, chainSpecificGasLimit uint32, err error) {
+func (b *BlockHistoryEstimator) BumpLegacyGas(originalGasPrice *assets.Wei, gasLimit uint32, maxGasPriceWei *assets.Wei, attempts []PriorAttempt) (bumpedGasPrice *assets.Wei, chainSpecificGasLimit uint32, err error) {
+	if b.config.BlockHistoryEstimatorCheckInclusionBlocks() > 0 {
+		if err = b.checkConnectivity(attempts); err != nil {
+			if errors.Is(err, ErrConnectivity) {
+				b.logger.Criticalw("Gas bumping is being prevented due to a detected connectivity issue; this requires immediate action to fix", "err", err)
+				promBlockHistoryEstimatorConnectivityFailureCount.WithLabelValues(b.chainID.String(), "legacy").Inc()
+			}
+			return nil, 0, err
+		}
+	}
 	return BumpLegacyGasPriceOnly(b.config, b.logger, b.getGasPrice(), originalGasPrice, gasLimit, maxGasPriceWei)
 }
 
-func (b *BlockHistoryEstimator) GetDynamicFee(gasLimit uint32, maxGasPriceWei *big.Int) (fee DynamicFee, chainSpecificGasLimit uint32, err error) {
+// checkConnectivity detects if the transaction is not being included due to
+// some kind of mempool propagation or connectivity issue rather than
+// insufficiently high pricing and returns error if so
+func (b *BlockHistoryEstimator) checkConnectivity(attempts []PriorAttempt) error {
+	percentile := int(b.config.BlockHistoryEstimatorCheckInclusionPercentile())
+	// how many blocks since broadcast?
+	latestBlockNum := b.getCurrentBlockNum()
+	if latestBlockNum == nil {
+		b.logger.Warn("Latest block is unknown; skipping inclusion check")
+		// can't determine anything if we don't have/know latest block num yet
+		return nil
+	}
+	expectInclusionWithinBlocks := int(b.config.BlockHistoryEstimatorCheckInclusionBlocks())
+	blockHistory := b.getBlocks()
+	if len(blockHistory) < expectInclusionWithinBlocks {
+		b.logger.Warnf("Block history in memory with length %d is insufficient to determine whether transaction should have been included within the past %d blocks", len(blockHistory), b.config.BlockHistoryEstimatorCheckInclusionBlocks())
+		return nil
+	}
+	for _, attempt := range attempts {
```

Comment:
Do we need to iterate over all attempts, or just choose the latest attempt assuming that has the highest bumped fees.